Flutter_Hero_Animation_Unmasked/Flutter_Hero_Animation_Unmasked_Part_1.md

```diff
@@ -0,0 +1,382 @@
+    } else {
+      element.visitChildren(_visitHero);
+    }
+  }
+
+  context.visitChildElements(_visitHero);
+  return heroes;
+}
+
+...
+
+@override
+void didPush(Route<dynamic> toRoute, Route<dynamic>? fromRoute) {
+    WidgetsBinding.instance?.addPostFrameCallback((Duration value) {
+        /// If the flight is not valid, let's just ignore the case
+        if (!_isFlightValid(fromRoute as PageRoute?, toRoute as PageRoute)) {
+          return;
+        }
+        final BuildContext fromContext = fromRoute!.subtreeContext!;
+        final BuildContext toContext = toRoute.subtreeContext!;
+
+        Map<String, UnmaskedHeroState> sourceHeroes = _inviteHeroes(fromContext);
+        for (UnmaskedHeroState hero in sourceHeroes.values) {
+        print(
+            "Source Hero invited: tag = ${hero.widget.tag}, type = ${hero.widget.child.runtimeType}");
+        }
+        Map<String, UnmaskedHeroState> destinationHeroes =
+            _inviteHeroes(toContext);
+        for (UnmaskedHeroState hero in destinationHeroes.values) {
+        print(
+            "Destination Hero invited: tag = ${hero.widget.tag}, type = ${hero.widget.child.runtimeType}");
+        }
+    });
+    super.didPush(toRoute, fromRoute);
+}
+```
+
+Let's break down and see what's going on here:
+
+1.  with the `_inviteHeroes` method, we recursively browse a given context to "visit" and find all instances of the `UnmaskedHero` widget.
+2.  in the `didPush` overriden method, we call the `_inviteHeroes` method on both **"from"** and **"to"** contexts so that we gather lists of source & destination heroes.
+3.  finally, we wrap this whole logic inside a callback passed to `WidgetsBinding.instance?.addPostFrameCallback`
+
+There are 2 interesting things, I would like to highlight:
+
+1. the `_visitHero` method shows how to navigate recursively through the Element tree to find some elements which are instances of widgets. If you want to learn more about Flutter's rendering behavior and the difference between Widget, Element & RenderObject, checkout [this talk on the subject](https://www.youtube.com/watch?v=996ZgFRENMs&t=1s)
+2. We call the `WidgetsBinding.instance?.addPostFrameCallback`. Because the `didPush` method of an observer is called right after the page is actually push into the navigation stack, the context of the destination route has not been mounted yet. Without `addPostFrameCallback`, the context of the `toRoute` page would be null and we would fallback into the case: `isFlightValid(...) == false`. With this callback, we wait for the newly pushed page to be built and ensure that the `toRoute.subtreeContext` is well defined, so that we can look for the instances of our Hero elements. If you want to understand more this method, take a look at this article: [addPostFrameCallback](https://www.didierboelens.com/fr/2019/04/addpostframecallback/) as well as the [documentation](https://api.flutter.dev/flutter/scheduler/SchedulerBinding/addPostFrameCallback.html).
```

Comment:
Pour l'article fr j'aurais rajouté peut être une petite indication `at this French article` peut être ?